### ISSUE TITLE
Children of Panels with clip content = true do not respect corner radii applied through panel styling

### ISSUE THREAD

**Issue by clarkejm** [godotengine/godot#18459]
**Godot version:**

3.0.2.stable.official

**OS/device including version:**

Win 10 Version 1709 build 16299.371
Lenovo P40 Yoga
OpenGL ES 3.0 Renderer: Quadro M500M/PCIe/SSE2

**Issue description:**

Intention was to use panel StyleBoxFlat to create rounded panels and use TextureRect as children to provide textured skins to each panel, however it appears the child nodes ignore the corners and are only clipped when leaving the margin of the panel.
![clipping](https://user-images.githubusercontent.com/24907845/39342867-a3960aca-499f-11e8-988e-58d898494081.PNG)

Please advise if there is a better way to set up textured and rounded panels.

**Steps to reproduce:**

Set up a panel
- Create a scene
- Add a Panel and size appropriately
- In Panel Properties>Rect set Clip Content = true
- Custom Styles>Panel  add New StyleBoxFlat
- Edit the Stylebox> set Corner Radii ~50 (so that they are easily visible)

Add a child sprite

- Add sprite as child of Panel (use godot icon.png for example)
- place sprite over the rounded corner, _note that it doesn't get clipped as it crosses the panel edge_ 

Add a TextureRect

- Add TextureRect as a child of panel
- size and position it to the same as the panel
- Apply icon.png as Texture and set Stretch Mode = Tile _note that it completely covers the panel_
- In TextureRect>CanvasItem>Visibility> set Show Behind Parent = true _note that the tiled texture can now bee seen poking out from behind the panel_

**Minimal reproduction project:**
[corner_clipping.zip](https://github.com/godotengine/godot/files/1953698/corner_clipping.zip)

**Comment by clarkejm** [comment]
Are there any better ways of applying a corner radius to a panel/control texture within Godot (i.e without changing the source image), currently StyleBoxTexture does not have corner radius properties which is what lead to this issue in the first place?

**Comment by groud** [comment]
No you have to change the texture. Or maybe use a shader

**Comment by clarkejm** [comment]
For anyone that's interested this is my hacky shader for applying corner radius and texture to a panel:

```
shader_type canvas_item;
render_mode unshaded;

uniform vec2 size; //rect_size of panel passed on_ready
uniform sampler2D TEXTURE2 : hint_normal; //Image texture to use as background
uniform vec2 texture_size; //rect size of texture (to scale texture)
uniform vec2 transform; // x,y transform to clip to areas of texture
uniform float radius : hint_range(0.0,0.5); // corner radius to apply to panel, will apply as percentage of shortest x/y length

void fragment()
{
    float yx_ratio = size.y/size.x;
    float xy_ratio = size.x/size.y;
    float grow = max(yx_ratio,xy_ratio);
    float shrink = min(yx_ratio,xy_ratio);
    ivec2 img_size = textureSize(TEXTURE2,0);
    float img_xratio = texture_size.x/size.x;
    float img_yratio = texture_size.y/size.y;
    float tfm_xratio = transform.x/size.x;
    float tfm_yratio = transform.y/size.y;
    COLOR.rgb = texture(TEXTURE2,vec2((UV.x/img_xratio)+tfm_xratio,(UV.y/img_yratio)+tfm_yratio)).rgb;
    vec2 top_left = vec2(radius,radius*shrink);
    vec2 top_right = vec2(1.0-radius,radius*shrink);
    vec2 bottom_left = vec2(radius,1.0-(radius*shrink));
    vec2 bottom_right = vec2(1.0-radius,1.0-(radius*shrink));
    
    if (UV.x <= top_left.x && UV.y <= top_left.y){
         if (!(distance(vec2(UV.x,UV.y*grow),vec2(top_left.x,top_left.y*grow))<radius)){
            COLOR.a = 0.0;
        }
    }
    if (UV.x>=top_right.x &&  UV.y <= top_right.y){
         if (!(distance(vec2(UV.x,UV.y*grow),vec2(top_right.x,top_right.y*grow))<radius)){
            COLOR.a = 0.0;
        }
    }
    if (UV.x<=bottom_left.x && UV.y>=bottom_left.y){
         if (!(distance(vec2(UV.x,UV.y*grow),vec2(bottom_left.x,bottom_left.y*grow))<radius)){
        COLOR.a = 0.0;
        }
    }
    if (UV.x>=bottom_right.x && UV.y>=bottom_right.y){
         if (!(distance(vec2(UV.x,UV.y*grow),vec2(bottom_right.x,bottom_right.y*grow))<radius)){
        COLOR.a = 0.0;
        }
    }
}
```

**Comment by pouleyKetchoupp** [comment]
I've just written a shader that can be used to add rounded corners to any ui element, without all the branching from the previous one (for optimization purpose).
Thanks @clarkejm for your shader, it helped a lot!

`radius` : amount of rounded corner from 0 (not rounded) to 1 (perfect circle)
`size` : optional parameter to handle different ratios, you can set the rect size from your ui element

It also handles basic parameters for texture and tint, but you can get rid of that part if you don't need it.

```
shader_type canvas_item;
render_mode blend_mix,unshaded;

uniform vec4 color : hint_color = vec4(1.0, 1.0, 1.0, 1.0);

uniform float radius : hint_range(0.0, 1.0) = 0.0;
uniform vec2 size = vec2(1.0, 1.0);

void fragment() {
	vec4 texture_color = texture(TEXTURE, UV);
	COLOR = texture_color * color;
	COLOR.a *= texture(mask_texture, UV).a;
	
	vec2 size_ratio = vec2(max(1.0, size.x / size.y), max(1.0, size.y / size.x));
	float half_radius = 0.5 * radius;
	vec2 dist_max = half_radius / size_ratio;
	vec2 edge_pos = clamp(UV, dist_max, 1.0 - dist_max);
	float edge_dist = distance(UV * size_ratio, edge_pos * size_ratio);
	COLOR.a *= step(edge_dist, half_radius + 0.000001);
}
```

**Comment by hacker1122** [comment]
> I've just written a shader that can be used to add rounded corners to any ui element, without all the branching from the previous one (for optimization purpose).
> Thanks @clarkejm for your shader, it helped a lot!
> 
> `radius` : amount of rounded corner from 0 (not rounded) to 1 (perfect circle)
> `size` : optional parameter to handle different ratios, you can set the rect size from your ui element
> 
> It also handles basic parameters for texture and tint, but you can get rid of that part if you don't need it.
> 
> ```
> shader_type canvas_item;
> render_mode blend_mix,unshaded;
> 
> uniform vec4 color : hint_color = vec4(1.0, 1.0, 1.0, 1.0);
> 
> uniform float radius : hint_range(0.0, 1.0) = 0.0;
> uniform vec2 size = vec2(1.0, 1.0);
> 
> void fragment() {
> 	vec4 texture_color = texture(TEXTURE, UV);
> 	COLOR = texture_color * color;
> 	
> 	vec2 size_ratio = vec2(max(1.0, size.x / size.y), max(1.0, size.y / size.x));
> 	float half_radius = 0.5 * radius;
> 	vec2 dist_max = half_radius / size_ratio;
> 	vec2 edge_pos = clamp(UV, dist_max, 1.0 - dist_max);
> 	float edge_dist = distance(UV * size_ratio, edge_pos * size_ratio);
> 	COLOR.a *= step(edge_dist, half_radius + 0.000001);
> }
> ```

Hi! I'm using your shader in my project. I need to ask you, is it possible to do the "squared corner" instead of the normal one? Like in this image
![photo_2019-05-18_18-28-42](https://user-images.githubusercontent.com/13876831/57973676-01528880-79ad-11e9-819f-4fc88ee00499.jpg)

**Comment by pouleyKetchoupp** [comment]
@hacker1122 Here's a version of the shader with squared corners.

For info, the only difference is how the distance with the edge is calculated.
Rounded corner: euclidian distance (standard one) -> sqrt(x²+y²)
Squared corner: manhattan distance -> abs(x) + abs(y)

```
shader_type canvas_item;
render_mode blend_mix,unshaded;

uniform vec4 color : hint_color = vec4(1.0, 1.0, 1.0, 1.0);

uniform float radius : hint_range(0.0, 1.0) = 0.0;
uniform vec2 size = vec2(1.0, 1.0);

void fragment() {
	vec4 texture_color = texture(TEXTURE, UV);
	COLOR = texture_color * color;
	
	vec2 size_ratio = vec2(max(1.0, size.x / size.y), max(1.0, size.y / size.x));
	float half_radius = 0.5 * radius;
	vec2 dist_max = half_radius / size_ratio;
	vec2 edge_pos = clamp(UV, dist_max, 1.0 - dist_max);
	vec2 edge_vec = UV * size_ratio - edge_pos * size_ratio;
	float edge_dist = abs(edge_vec.x) + abs(edge_vec.y);
	COLOR.a *= step(edge_dist, half_radius + 0.000001);
}
```

**Comment by Calinou** [comment]
Closing, as StyleBoxFlat corners aren't meant to clip their children.

In Godot 4.0, you can use CanvasItem clipping to clip children automatically based on the parent's alpha channel.

**Comment by biblical-text** [comment]
I tried these shaders. The angle of the corners (curved or straight version) alters depending on the shape of the node you are shading.  If the node is square, you get perfect corners. If the node is a rectangle, the curve of the corners slants and looks weird looking. So basically it only works if you want to shade a square, or you want to fully curve something to be a complete circle. It doesn't shade like a normal "corner radius" type function where the corners all look evenly rounded.

<img width="127" alt="image" src="https://user-images.githubusercontent.com/88878602/213947630-b2ec5842-6f2d-412f-9d05-9ec867a8bf5c.png">


I spent an hour trying to work out the shader code to see if I can make it do rounded corners, but I was not able to work it out. Does anyone know how to make this shader create true rounded corners?
